File: APM/Java/Interpreter/src/ro/sandorrobertk94/domain/MyStack.java
package ro.sandorrobertk94.domain;

import java.util.EmptyStackException;

/**
 * Created by robert on 10/25/15.
 */
public class MyStack implements MyIStack {
    private static final int CAPACITY = 50;
    private int top;
    private Object[] stackArray;

    public MyStack() {
        top = 0;
        stackArray = new Object[CAPACITY];
    }

    @Override
    public int size() {
        return top;
    }

    @Override
    public boolean isEmpty() {
        return top == 0;
    }

    @Override
    public Object top() throws EmptyStackException {
        if (isEmpty()) {
            throw new EmptyStackException();
        }

        return stackArray[top - 1];
    }

    @Override
    public void push(Object o) {
        stackArray[top] = o;
        top++;
    }

    @Override
    public Object pop() throws EmptyStackException {
        if (isEmpty()) {
            throw new EmptyStackException();
        }

        top--;
        Object o = stackArray[top];
        stackArray[top] = null;
        return o;
    }

    @Override
    public String toString() {
        String out = "( \n";
        for (int i = top-1; i >= 0; i--) {
            out += "\t" + stackArray[i].toString() + "\n";
        }
        out += ");";
        return out;
    }
}

File: APM/CSharp/Interpreter/Interpreter/domain/MyStack.cs
namespace Interpreter.domain
{
    public class MyStack : MyIStack
    {
        private const int CAPACITY = 50;
        private readonly object[] stackArray;
        private int top;

        public MyStack()
        {
            stackArray = new object[CAPACITY];
            top = 0;
        }

        public bool IsEmpty()
        {
            return top == 0;
        }

        public object Top()
        {
            return IsEmpty() ? null : stackArray[top - 1];
        }

        public void Push(object obj)
        {
            stackArray[top] = obj;
            top++;
        }

        public object Pop()
        {
            if (IsEmpty())
            {
                return null;
            }

            var o = Top();
            top--;
            stackArray[top] = null;
            return o;
        }

        public override string ToString()
        {
            var o = "( \n";
            for (var i = top - 1; i >= 0; i--)
            {
                o += "\t" + stackArray[i] + "\n";
            }
            o += ");";
            return o;
        }
    }
}